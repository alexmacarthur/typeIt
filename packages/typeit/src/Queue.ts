import asArray from "./helpers/asArray";
import { QueueItem } from "./types";

const Queue = function (initialItems: QueueItem[]) {
  /**
   * Add a single or several steps onto the `waiting` queue.
   */
  const add = function (steps: QueueItem[] | QueueItem): typeof Queue {
    _queue = _queue.concat(asArray<QueueItem>(steps));

    return this;
  };

  /**
   * Given an index, set an item in the queue.
   */
  const set = function (index: number, item: QueueItem): void {
    _queue[index] = item;
  };

  /**
   * Move all `executed` queue items to `waiting`.
   */
  const reset = function (): void {
    _queue = _queue.map((item) => {
      delete item.done;

      return item;
    });
  };

  /**
   * Retrieve all items that are still eligible to be executed.
   */
  const getItems = (): QueueItem[] => _queue.filter(i => !i.done);

  const markDone = (index: number) => {
    _queue[index].done = true;
  }

  let _queue: QueueItem[] = [];

  add(initialItems);

  return {
    add,
    set,
    reset,
    getItems,
    markDone,
  };
};

export default Queue;
