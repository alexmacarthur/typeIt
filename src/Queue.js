import guaranteeThreeKeys from "./helpers/guaranteeThreeKeys";
import queueMany from "./helpers/queueMany";

export default function Queue(initialItem) {
  /**
   * Insert items into the `waiting` queue.
   * @todo Can we part with this?
   *
   * @param {integer} start
   * @param {array} newItems
   */
  this.insert = function(start, newItems) {
    _queue.splice(start, 0, newItems);
  };

  /**
   * Add a single or several steps onto the `waiting` queue.
   *
   * @param {array} stepOrSteps
   * @param {boolean} toBeginning
   * @return {object}
   */
  this.add = function(stepOrSteps, numberOfTimes, toBeginning) {
    // If a single thing is passed, assume it's an action with no argument.
    stepOrSteps = Array.isArray(stepOrSteps)
      ? stepOrSteps
      : [stepOrSteps, null];
    toBeginning = toBeginning || false;
    numberOfTimes = numberOfTimes || 1;
    let isMultipleSteps = Array.isArray(stepOrSteps[0]);

    if (!isMultipleSteps) {
      stepOrSteps = queueMany(numberOfTimes, stepOrSteps);
    }

    _queue = toBeginning
      ? stepOrSteps.concat(_queue)
      : _queue.concat(stepOrSteps);

    return this;
  };

  /**
   * Given an index, remove the last item on the `waiting` queue.
   *
   * @param {integer} index
   * @return {object}
   */
  this.delete = function(index) {
    _queue.splice(index, 1);
    return this;
  };

  /**
   * Move all `executed` queue items to `waiting`.
   *
   * @return {object}
   */
  this.reset = function() {
    _queue = guaranteeThreeKeys(_queue).map(item => {
      item[2].executed = false;
      return item;
    });

    return this;
  };

  /**
   * Retrieve all items that are still eligible to be executed.
   *
   * @return {array}
   */
  this.getItems = function() {
    return guaranteeThreeKeys(_queue).filter(i => !i.executed);
  };

  let _queue = [];

  // Don't include initial item if we're recycling
  // items from a previous run.
  if (initialItem) {
    this.add(initialItem);
  }
}
