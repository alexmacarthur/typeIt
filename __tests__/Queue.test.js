import Queue from "../src/Queue.js";

let queue;

beforeEach(() => {
  queue = new Queue();
});

describe("add()", () => {
  test("It should add steps properly.", () => {
    expect(queue.getItems()).toEqual([]);

    queue.add(["step1"]);
    queue.add(["step2"]);

    expect(queue.getItems()).toMatchSnapshot();
  });

  test("It should add a step passed as a function.", () => {
    const myFunc = () => "func!";
    queue.add(myFunc);
    queue.add(myFunc);
    queue.add(myFunc);
    expect(queue.getItems()).toMatchSnapshot();
  });

  test("It should add multiple steps passed at once.", () => {
    const myFunc = () => "func!";
    queue.add([
      [myFunc, 1],
      [myFunc, 2],
      [myFunc, 3]
    ]);
    expect(queue.getItems()).toMatchSnapshot();
  });
});

test("It should reset properly by marking each item as having not yet been executed.", () => {
  queue.add([
    ["value1", null, { executed: true }],
    ["value2", null, { executed: true }],
    ["value3", null, { executed: false }],
    ["value4", null, { executed: true }],
    ["value5", null, { executed: true }]
  ]);

  queue.reset();

  expect(queue.getItems()).toMatchSnapshot();
});

test("It should set initial steps properly.", () => {
  let q1 = new Queue();

  expect(q1.getItems()).toEqual([]);

  let items = [
    ["value1", null, {}],
    ["value2", null, {}],
    ["value3", null, {}]
  ];

  let q2 = new Queue(items);

  expect(q2.getItems()).toMatchSnapshot();
});
