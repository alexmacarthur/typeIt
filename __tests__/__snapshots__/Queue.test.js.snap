// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should only return non-executed items. 1`] = `
Array [
  Array [
    "value2",
    null,
    Object {
      "executed": false,
    },
  ],
  Array [
    "value3",
    null,
    Object {},
  ],
]
`;

exports[`It should reset properly by marking each item as having not yet been executed. 1`] = `
Array [
  Array [
    "first-action",
    null,
    Object {
      "executed": false,
    },
  ],
  Array [
    "value1",
    null,
    Object {
      "executed": false,
    },
  ],
  Array [
    "value2",
    null,
    Object {
      "executed": false,
    },
  ],
  Array [
    "value3",
    null,
    Object {
      "executed": false,
    },
  ],
  Array [
    "value4",
    null,
    Object {
      "executed": false,
    },
  ],
  Array [
    "value5",
    null,
    Object {
      "executed": false,
    },
  ],
]
`;

exports[`It should set initial steps properly. 1`] = `
Array [
  Array [
    "value1",
    null,
    Object {},
  ],
  Array [
    "value2",
    null,
    Object {},
  ],
  Array [
    "value3",
    null,
    Object {},
  ],
]
`;

exports[`add() It should add multiple steps passed at once. 1`] = `
Array [
  Array [
    "first-action",
    null,
    Object {},
  ],
  Array [
    [Function],
    1,
    Object {},
  ],
  Array [
    [Function],
    2,
    Object {},
  ],
  Array [
    [Function],
    3,
    Object {},
  ],
]
`;

exports[`add() It should add steps properly. 1`] = `
Array [
  Array [
    "first-action",
    null,
    Object {},
  ],
  Array [
    "step1",
    null,
    Object {},
  ],
  Array [
    "step2",
    null,
    Object {},
  ],
]
`;
